agent/local/state.go

```diff
@@ -0,0 +1,1031 @@
+func (l *State) Service(id string) *structs.NodeService {
+	l.RLock()
+	defer l.RUnlock()
+
+	s := l.services[id]
+	if s == nil || s.Deleted {
+		return nil
+	}
+	return s.Service
+}
+
+// Services returns the locally registered services that the
+// agent is aware of and are being kept in sync with the server
+func (l *State) Services() map[string]*structs.NodeService {
+	l.RLock()
+	defer l.RUnlock()
+
+	m := make(map[string]*structs.NodeService)
+	for id, s := range l.services {
+		if s.Deleted {
+			continue
+		}
+		m[id] = s.Service
+	}
+	return m
+}
+
+// ServiceState returns a shallow copy of the current service state
+// record. The service record still points to the original service
+// record and must not be modified.
+func (l *State) ServiceState(id string) *ServiceState {
+	l.RLock()
+	defer l.RUnlock()
+
+	s := l.services[id]
+	if s == nil || s.Deleted {
+		return nil
+	}
+	return s.Clone()
+}
+
+// SetServiceState is used to overwrite a raw service state with the given
+// state. This method is safe to be called concurrently but should only be used
+// during testing. You should most likely call AddService instead.
+func (l *State) SetServiceState(s *ServiceState) {
+	l.Lock()
+	defer l.Unlock()
+
+	l.services[s.Service.ID] = s
+	l.TriggerSyncChanges()
+}
+
+// ServiceStates returns a shallow copy of all service state records.
+// The service record still points to the original service record and
+// must not be modified.
+func (l *State) ServiceStates() map[string]*ServiceState {
+	l.RLock()
+	defer l.RUnlock()
+
+	m := make(map[string]*ServiceState)
+	for id, s := range l.services {
+		if s.Deleted {
+			continue
+		}
+		m[id] = s.Clone()
+	}
+	return m
+}
+
+// CheckToken is used to return the configured health check token for a
+// Check, or if none is configured, the default agent ACL token.
+func (l *State) CheckToken(checkID types.CheckID) string {
+	l.RLock()
+	defer l.RUnlock()
+	return l.checkToken(checkID)
+}
+
+// checkToken returns an ACL token associated with a check.
+// This method is not synchronized and the lock must already be held.
+func (l *State) checkToken(id types.CheckID) string {
+	var token string
+	c := l.checks[id]
+	if c != nil {
+		token = c.Token
+	}
+	if token == "" {
+		token = l.tokens.UserToken()
+	}
+	return token
+}
+
+// AddCheck is used to add a health check to the local state.
+// This entry is persistent and the agent will make a best effort to
+// ensure it is registered
+func (l *State) AddCheck(check *structs.HealthCheck, token string) error {
+	if check == nil {
+		return fmt.Errorf("no check")
+	}
+
+	if l.discardCheckOutput.Load().(bool) {
+		check.Output = ""
+	}
+
+	// if there is a serviceID associated with the check, make sure it exists before adding it
+	// NOTE - This logic may be moved to be handled within the Agent's Addcheck method after a refactor
+	if check.ServiceID != "" && l.services[check.ServiceID] == nil {
```

Comment:
You need to be holding the lock in order to read `l.services` here.

Reply:
Indeed